File: from-1901-to-2000/1971-find-if-path-exists-in-graph/solution.go
func validPath(n int, edges [][]int, source int, destination int) bool {
    adj := make(map[int][]int)
    for _, e := range edges {
        adj[e[0]] = append(adj[e[0]], e[1])
        adj[e[1]] = append(adj[e[1]], e[0])
    }
    seen := make(map[int]bool)
    q := []int{source}

    for len(q) > 0 {
        v := q[0]
        if v == destination {
            return true
        }
        q = q[1:]
        if seen[v] {
            continue
        }
        seen[v] = true
        for _, a := range adj[v] {
            q = append(q, a)
        }
    }
    return false
}

File: from-1901-to-2000/1971-find-if-path-exists-in-graph/solution.py
class Solution:
    def validPath(self, n: int, edges: List[List[int]], source: int, destination: int) -> bool:
        if n == 1:
            return True
            
        g = defaultdict(set)
        for e in edges:
            g[e[0]].add(e[1])
            g[e[1]].add(e[0])

        visited = set()
        q = deque()
        q.append(source)

        while q:
            node = q.popleft()
            visited.add(node)
            for nbr in g[node]:
                if nbr == destination:
                    return True
                if nbr not in visited:
                    q.append(nbr)

        return False
